### ISSUE TITLE
include-what-you-use suggests removal of header which declares a namespace

### ISSUE THREAD

**Issue by pseyfert** [include-what-you-use/include-what-you-use#441]
Hi,
running IWYU on [lhcb/LHCb](https://gitlab.cern.ch/lhcb/LHCb) I encountered a few wrong suggestions which I don't understand yet and I'd like to figure out what's wrong without just putting pragmas all over the code. I tried to put a stripped down version on github:

[one failing file example](https://github.com/pseyfert/iwyu-failure-on-lhcbmath)

(commands I ran for compilation are in the build directory, I used clang 3.9 and iwyu from the `clang_3.9` branch). It basically is a header file `Package/file.h` and an implementation file `src/file.cpp`. The header declares
```
namespace Math {
  unsigned long long choose(const unsigned short n, const unsigned short k);
}
```
and the implementation file has
```
#include "Package/file.h"
unsigned long long Math::choose(const unsigned short n, const unsigned short k) {…}
```
IWYU suggests to remove the inclusion of `Package/file.h` (which then leads to a compiler error since the namespace `Math` hasn't been declared yet). w/o using the namespace at all, the suggestion can be implemented.

That said, there has to be something else relevant I'm missing, because I haven't reproduced the wrong suggestion in a more minimal example yet.

Thanks in advance for suggestions and pointers.

**Comment by vsapsai** [comment]
Thanks for the report, I was able to reproduce it with

```c++
// issue441-d1.h
namespace ns {
  void foo();
}

// issue441.cc
#include "issue441-d1.h"

void ns::foo() {}
```

It is important that .h and .cc files have different names. Probably that's why you had trouble creating a smaller repro case.

**Comment by pseyfert** [comment]
Thanks for checking, however the stems of the filenames are the same (Choose.h, Choose.cpp), they are just in different directories and the include statement contains that:
```
../Kernel/LHCbMath/src/Choose.cpp should remove these lines:
- #include "LHCbMath/Choose.h"  // lines 14-14
```

and I'm not able to reproduce with the same-stem-different-directory setup:
```
// issue/issue441.h
namespace ns {
  void foo();
}

// issue441.cc
#include "issue/issue441.h"

void ns::foo() {}

//outputs:
//(issue/issue441.h has correct #includes/fwd-decls)
//
//(issue441.cc has correct #includes/fwd-decls)
//
```

**Comment by kimgr** [comment]
The associated-header game has a few more rules than that, see:
https://github.com/include-what-you-use/include-what-you-use/blob/23253ec2e8c6b84a9b1a45329b7a1c1a81233fd8/iwyu_preprocessor.cc#L995-L1019

I think your LHCbMath example has other includes before the associated header, and therefore it's not recognized. Your minimal example above only has the one include with the same stem.

Anyway, the problem still holds; it looks like we're incorrectly discarding headers defining namespaces we depend on.

**Comment by pseyfert** [comment]
For using it in LHCb, we now went for https://github.com/pseyfert/include-what-you-use/commit/fa77858457fe1563f2bcc85fc38ff95ad7de61c1 (i.e. assume the associated header has the same base filename and ignore the directory). Not the best fix imaginable but good enough for us for now.

**Comment by Tim020** [comment]
Has there been any progress on getting this issue resolved? I am seeing the same sort of thing, removing an include which declares a namespace

**Comment by kimgr** [comment]
@Tim020 If the .cc and header are named wildly differently, as here, you might be able to use the associated pragma in your .cc file:

    // strnstr.cc:
    #include "platform.h"  // IWYU pragma: associated

**Comment by pseyfert** [comment]
@Tim020 I'm still using the hack I posted above, with a few more patches I cherry-picked from master. (It's not fully bug-free as it assumes we have no two headers with the same name in different directories - there are such files, luckily I haven't yet observed that `foo/problem.h` and `baz/problem.h` get both included in a file `lorem/problem.cc`, which saved me so far from real trouble).

I went for this hack as it seemed more practical to me than spamming the project with pragmas or reordering includes everywhere, and I don't think I'll revisit this choice soon.
